### PR TITLE
Added ansible role syncing and delete ansible role...

### PR DESCRIPTION
**Pull request by amolpati30**
1.  While importing the Ansible role, it was noticed that the 'RedHatInsights.insights-client' role had already been imported. This caused issues with the Ansible test cases, resulting in errors due to conflicts between the synced role and the imported role. To resolve this, I used the sync option, which automatically syncs the selected role without requiring the import of all roles. 

2.  The `add_single_ansible_role` method will now accept an additional parameter, `role`, which will determine the Ansible role to be added to the host.

Dependent PR: https://github.com/SatelliteQE/airgun/pull/1542

### CODE REVIEW

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors'

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors'

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8256
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors --external-logging
Test Result : ========= 2 passed, 12 deselected, 145 warnings in 1298.81s (0:21:38) ==========
```

File: tests/foreman/ui/test_ansible.py [hunk 1]
```diff
@@ -517,6 +514,7 @@ def test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors(
         module_target_sat.api.AnsibleRoles().sync(
             data={'proxy_id': nc.id, 'role_names': SELECTED_ROLE}
         )
+        module_target_sat.cli.Ansible.roles_delete({'name': 'RedHatInsights.insights-client'})
```
[the remark this answers is not in the document]
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:517]: Yes, the role was not being used, but I removed it intentionally because it was being imported automatically before the test started. This was causing issues when adding the Ansible role, so I decided to remove it

[the remark this answers is not in the document]
Comment (shweta83) [reply tests/foreman/ui/test_ansible.py:517]: Can you please delete the role in the test from where it was imported? You can add teardown in that test.

[the remark this answers is not in the document]
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:517]: I attempted to use the teardown, but it didn't work here.

[the remark this answers is not in the document]
Comment (shweta83) [reply tests/foreman/ui/test_ansible.py:517]: You wouldn't need to delete any role now.  You can skip this step. It works well.

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors'

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8332
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors --external-logging
Test Result : ========= 2 passed, 12 deselected, 145 warnings in 1339.19s (0:22:19) ==========
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py -k 'TestAnsibleCfgMgmt'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8410
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py -k TestAnsibleCfgMgmt --external-logging
Test Result : = 3 failed, 5 passed, 15 deselected, 232 warnings, 1 error in 1568.66s (0:26:08) =
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_assign_and_remove_ansible_role_to_host or test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_non_admin_user_access_with_usergroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8501
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_assign_and_remove_ansible_role_to_host or test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_non_admin_user_access_with_usergroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors --external-logging
Test Result : ========= 4 passed, 10 deselected, 283 warnings in 2488.88s (0:41:28) ==========
```

File: tests/foreman/ui/test_ansible.py [hunk 2]
```diff
@@ -471,15 +468,14 @@ def test_positive_non_admin_user_access_with_usergroup(
         user = target_sat.api.User(
             login=gen_string('alpha'), password=password, admin=False
         ).create()
-        request.addfinalizer(user.delete)
```
Comment (Gauravtalreja1) [inline tests/foreman/ui/test_ansible.py:474]: why removing user.delete from finalizer, this looks unrelated?

Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:474]: The issue is causing conflicts and leading to test failures. After removing this, the tests are running successfully

Comment (Gauravtalreja1) [reply tests/foreman/ui/test_ansible.py:474]: what conflict do you see? its different entity, not sure what conflict it may cause

Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:474]: Screenshot from 2024-09-11 00-48-40

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_assign_and_remove_ansible_role_to_host or test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_non_admin_user_access_with_usergroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8639
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_assign_and_remove_ansible_role_to_host or test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_non_admin_user_access_with_usergroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors --external-logging
Test Result : ===== 4 passed, 10 deselected, 184 warnings, 1 error in 1515.79s (0:25:15) =====
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_non_admin_user_access_with_usergroup'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8641
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_non_admin_user_access_with_usergroup --external-logging
Test Result : ====== 1 passed, 13 deselected, 24 warnings, 1 error in 743.76s (0:12:23) ======
```

File: tests/foreman/ui/test_ansible.py (hunk 2 above)
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:474]: Getting this error - `failed on teardown with "requests.exceptions.HTTPError: 422 Client Error: Unprocessable Content for url: https://ip-10-0-216-32.rhos-01.prod.psi.rdu2.redhat.com/api/v2/users/5"`
Any other solution for this?

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_non_admin_user_access_with_usergroup'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8644
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_non_admin_user_access_with_usergroup --external-logging
Test Result : ====== 1 passed, 13 deselected, 24 warnings, 1 error in 765.28s (0:12:45) ======
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_non_admin_user_access_with_usergroup'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8653
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_non_admin_user_access_with_usergroup --external-logging
Test Result : ====== 1 passed, 13 deselected, 24 warnings, 1 error in 731.29s (0:12:11) ======
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_non_admin_user_access_with_usergroup'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8659
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_non_admin_user_access_with_usergroup --external-logging
Test Result : ========== 1 failed, 13 deselected, 24 warnings in 791.38s (0:13:11) ===========
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_non_admin_user_access_with_usergroup'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8662
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_non_admin_user_access_with_usergroup --external-logging
Test Result : ========== 1 failed, 13 deselected, 24 warnings in 762.21s (0:12:42) ===========
```

File: tests/foreman/ui/test_ansible.py (hunk 2 above)
Comment (shweta83) [reply tests/foreman/ui/test_ansible.py:474]: @Gauravtalreja1 What is the reason to not delete the user in teardown?

File: tests/foreman/ui/test_ansible.py [hunk 3]
```diff
@@ -467,22 +467,23 @@ def test_positive_non_admin_user_access_with_usergroup(
...
+        id = target_sat.nailgun_smart_proxy.id
+        target_sat.api.AnsibleRoles().sync(data={'proxy_id': id, 'role_names': [SELECTED_ROLE]})
+        with target_sat.ui_session(test_name, user=user.login, password=password) as session:
+            session.host_new.add_single_ansible_role(target_sat.hostname, SELECTED_ROLE)
             wait_for(lambda: session.browser.refresh(), timeout=5)
             ansible_roles_table = session.host_new.get_ansible_roles(target_sat.hostname)
             assert ansible_roles_table[0]['Name'] == SELECTED_ROLE
+        with target_sat.ui_session() as session:
+            session.user.delete(name)
+            assert not session.user.search(name)
```
Comment (shweta83) [inline tests/foreman/ui/test_ansible.py:486]: We can use api to delete the user or add finalizer to do it.

Comment (shweta83) [reply tests/foreman/ui/test_ansible.py:486]: If the teardown is causing conflict then I would suggest to try the following approach
```
session.host_new.remove_single_ansible_role(target_sat.hostname)
 session.login.logout()
```

and use teardown to delete the user.

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_non_admin_user_access_with_usergroup'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8701
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_non_admin_user_access_with_usergroup --external-logging
Test Result : ====== 1 passed, 13 deselected, 25 warnings, 1 error in 765.49s (0:12:45) ======
```

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8702
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_non_admin_user_access_with_usergroup --external-logging
Test Result : ====== 1 passed, 13 deselected, 25 warnings, 1 error in 817.17s (0:13:37) ======
```

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_non_admin_user_access_with_usergroup'
airgun: 1542

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 8704
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k test_positive_non_admin_user_access_with_usergroup --external-logging
Test Result : ========== 1 passed, 13 deselected, 24 warnings in 827.30s (0:13:47) ===========
```

File: tests/foreman/ui/test_ansible.py [hunk 4]
```diff
@@ -449,39 +446,48 @@ def test_positive_assign_and_remove_ansible_role_to_hostgroup(
...
         ).create()
-        assert user.login in [u.read().login for u in user_gp.user]
-        with target_sat.ui_session(test_name, user.login, password) as session:
-            id = target_sat.nailgun_smart_proxy.id
-            target_sat.api.AnsibleRoles().sync(data={'proxy_id': id, 'role_names': [SELECTED_ROLE]})
-            session.host_new.add_single_ansible_role(target_sat.hostname)
+        assert non_a_user.login in [u.read().login for u in user_gp.user]
+        id = target_sat.nailgun_smart_proxy.id
+        target_sat.api.AnsibleRoles().sync(data={'proxy_id': id, 'role_names': [SELECTED_ROLE]})
+        with target_sat.ui_session(test_name, user=non_a_user.login, password=password) as session:
```
Comment (shweta83) [inline tests/foreman/ui/test_ansible.py:487]: This should be done in some organization, can be module_org.

File: tests/foreman/ui/test_ansible.py [hunk 5]
```diff
@@ -449,39 +446,48 @@ def test_positive_assign_and_remove_ansible_role_to_hostgroup(
...
+            session.host_new.add_single_ansible_role(host.name, SELECTED_ROLE)
             wait_for(lambda: session.browser.refresh(), timeout=5)
-            ansible_roles_table = session.host_new.get_ansible_roles(target_sat.hostname)
+            ansible_roles_table = session.host_new.get_ansible_roles(host.name)
```
Comment (shweta83) [inline tests/foreman/ui/test_ansible.py:490]: This is also an additional step which is not part of the original bug.

**Review by shweta83 — CHANGES_REQUESTED** [review]

File: tests/foreman/ui/test_ansible.py (hunk 5 above)
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:490]: I added this for assertion purposely to verify whether the role has been assigned

File: tests/foreman/ui/test_ansible.py (hunk 4 above)
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:487]: In my opinion, this is not required. The next task can be performed on any organization; only a host is needed

File: tests/foreman/ui/test_ansible.py (hunk 2 above)
Comment (Gauravtalreja1) [reply tests/foreman/ui/test_ansible.py:474]: I think this question is for @amolpati30

File: tests/foreman/ui/test_ansible.py [hunk 6]
```diff
@@ -449,39 +446,48 @@ def test_positive_assign_and_remove_ansible_role_to_hostgroup(
...
+        :Verifies: SAT-15826
 
         :customerscenario: true
         """
         SELECTED_ROLE = 'RedHatInsights.insights-client'
+        name = gen_string('alpha')
         password = gen_string('alpha')
-        user = target_sat.api.User(
-            login=gen_string('alpha'), password=password, admin=False
+        admin_user = target_sat.api.User(
```
[the remark this answers is not in the document]
Comment (shweta83) [reply tests/foreman/ui/test_ansible.py:475]: You can remove the role from user and logout the session. I am guessing that an open user session is causing the conflict in the teardown. After that, you should be able to delete the user.

File: tests/foreman/ui/test_ansible.py (hunk 2 above)
Comment (shweta83) [reply tests/foreman/ui/test_ansible.py:474]: @amolpati30 Deleting the user in teardown should work.

File: tests/foreman/ui/test_ansible.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:475]: https://github.com/SatelliteQE/robottelo/commit/919bb87a74f566b052cca7dae17be39e2723d2a5 I tried your suggestion, but that wasn't the issue. When I log in as a non-admin user and the owner is not set to Satellite, assigning a role to the sat causes the non-admin user to become the owner, even after I remove the role. I am still facing the same teardown issue, even after logging out. To prevent this, I created the host and assigned the admin as the owner, so when the non-admin user logs in and performs certain actions, it won’t affect the teardown process later.

File: tests/foreman/ui/test_ansible.py (hunk 2 above)
Comment (Gauravtalreja1) [reply tests/foreman/ui/test_ansible.py:474]: @amolpati30 so could you add the finalizer back?

File: tests/foreman/ui/test_ansible.py [hunk 7]
```diff
@@ -449,39 +446,48 @@ def test_positive_assign_and_remove_ansible_role_to_hostgroup(
...
-        :expectedresults: See Ansible page as user should be administrator due to usergroup
+        :Verifies: SAT-15826
 
         :customerscenario: true
         """
         SELECTED_ROLE = 'RedHatInsights.insights-client'
+        name = gen_string('alpha')
         password = gen_string('alpha')
-        user = target_sat.api.User(
-            login=gen_string('alpha'), password=password, admin=False
```
Comment (Gauravtalreja1) [inline tests/foreman/ui/test_ansible.py:472]: admin user exists as default, so why creating a new user?

**Comment by amolpati30** [comment]
trigger: test-robottelo
pytest: tests/foreman/ui/test_ansible.py::TestAnsibleCfgMgmt -k 'test_positive_assign_and_remove_ansible_role_to_hostgroup or test_positive_ansible_config_report_changes_notice_and_failed_tasks_errors or test_positive_assign_and_remove_ansible_role_to_host'

File: tests/foreman/ui/test_ansible.py (hunk 7 above)
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:472]: Removed

File: tests/foreman/ui/test_ansible.py (hunk 2 above)
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py:474]: I have re-added the finalizer, made some adjustments, and it is now working.

File: tests/foreman/ui/test_ansible.py (hunk 4 above)
Comment (amolpati30) [reply tests/foreman/ui/test_ansible.py]: Thank you for your suggestion. I have added both the location and organization

**Review by Gauravtalreja1 — APPROVED** [review]
